moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp

```diff
@@ -797,31 +809,52 @@ bool IKFastKinematicsPlugin::searchPositionIK(const geometry_msgs::Pose& ik_pose
   {
     ROS_DEBUG_STREAM_NAMED("ikfast", "No need to search since no free params/redundant joints");
 
-    // Find first IK solution, within joint limits
-    if (!getPositionIK(ik_pose, ik_seed_state, solution, error_code))
+    std::vector<geometry_msgs::Pose> ik_poses(1, ik_pose);
+    std::vector<std::vector<double>> solutions;
+    kinematics::KinematicsResult kinematic_result;
+    // Find all IK solution within joint limits
+    if (!getPositionIK(ik_poses, ik_seed_state, solutions, kinematic_result, options))
     {
       ROS_DEBUG_STREAM_NAMED("ikfast", "No solution whatsoever");
       error_code.val = moveit_msgs::MoveItErrorCodes::NO_IK_SOLUTION;
       return false;
     }
 
-    // check for collisions if a callback is provided
-    if (!solution_callback.empty())
+    // sort solutions by their distance to the seed
+    std::vector<LimitObeyingSol> solutions_obey_limits;
+    for (std::size_t i = 0; i < solutions.size(); ++i)
     {
-      solution_callback(ik_pose, solution, error_code);
-      if (error_code.val == moveit_msgs::MoveItErrorCodes::SUCCESS)
+      double dist_from_seed = 0.0;
+      for (std::size_t j = 0; j < ik_seed_state.size(); ++j)
       {
-        ROS_DEBUG_STREAM_NAMED("ikfast", "Solution passes callback");
-        return true;
+        dist_from_seed += fabs(ik_seed_state[j] - solutions[i][j]);
       }
-      else
+
+      solutions_obey_limits.push_back({ solutions[i], dist_from_seed });
+    }
+    std::sort(solutions_obey_limits.begin(), solutions_obey_limits.end());
+
+    // check for collisions if a callback is provided
+    if (!solution_callback.empty())
+    {
+      for (std::size_t i = 0; i < solutions_obey_limits.size(); ++i)
       {
-        ROS_DEBUG_STREAM_NAMED("ikfast", "Solution has error code " << error_code);
-        return false;
+        solution_callback(ik_pose, solutions_obey_limits[i].value, error_code);
```

Comment:
Currently you have one for-loop for computing the distance to the seed and another one for collision checks.  Would it not be more efficient to have a single for-loop that checks for collisions first and adds the solution to `solutions_obey_limits` only if it passes the collision test?